<!DOCTYPE HTML>
<!--
	Projection by TEMPLATED
	templated.co @templatedco
	Released for free under the Creative Commons Attribution 3.0 license (templated.co/license)
-->
<html>
	<head>
		<title>Generic - Projection by TEMPLATED</title>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1" />
		<link rel="stylesheet" href="assets/css/main.css" />
	</head>
	<body class="subpage">

		<!-- Header -->
			<header id="header">
				<div class="inner">
					<a href="https://www.linode.com/" class="logo">Hosted by Linode</a>
					<nav id="nav">
						<a href="https://github.com/SymPortal/SymPortal_framework">source code</a>
						<a href="https://github.com/SymPortal/SymPortal_framework/wiki/Welcome-to-SymPortal">wiki</a>
						<a href="index_templated.html">home</a>
					</nav>
					<a href="#navPanel" class="navPanelToggle"><span class="fa fa-bars"></span></a>
				</div>
			</header>

		<!-- Three -->
			<section id="three" class="wrapper">
				<div class="inner">
					<header class="align-center">
						<h2>Submitting data to SymPortal</h2>
						<p><font color="#565656">Why submit my data to SymPortal and how is my data used?</font></p>
					</header>
					<div class="image round left">
						<img src="images/text_two.jpg" alt="Pic 01" />
					</div>
					<p><font color="#565656">Why submit my data?</font><br/>Within the SymPortal analysis, a central
						principle is applied that enables SymPortal to distinguish between intra- and inter-genomic
						sources of ITS2 sequence variants. That principle is: the probability that a given set of
						ITS2 sequences found in a single coral sample are representative of a single Symbiodiniaceae
						genotype increases with the number of samples that set is found in. This principle is not new
						to SymPortal but rather has been used in DGGE-based methodologies for over 15 years. These
						gel-based methodologies have been able to successfully identify genetically distinct
						Symbiodiniaceae taxa that have been verified by additional genetic markers.	So what does this
						have to do with my data?
					</p>
					<p>Let’s look at an example. Let’s say you
						have 96 samples in your dataset, and each of the samples in your dataset contains the same
						Symbiodiniaceae taxa. When run through an analysis, SymPortal will identify the same reoccurring
						sets of ITS2 sequences in each of your samples. These sequences will be identified as defining
						intragenomic variants (DIVs) and used to define an ITS2 type profile, e.g. C3-C3a-C3cc.
						SymPortal is predicting that this set of sequences is representative of a Symbiodiniaceae taxon.
						It makes this prediction based on the central principle explained above. Essentially, because
						your dataset contains a sufficient number of samples, with the same reoccurring sets of
						sequences, you enable SymPortal to make this prediction. In contrast, let’s consider a case
						where your dataset contains Symbiodiniaceae taxa that only occur in a low number of your
						samples. In this case, SymPortal will be limited in its ability to resolve these taxa.
						For example, if two of your samples contain a Symbiodiniaceae taxon not found in other
						samples of your dataset, SymPortal will not be able to predict an accurate ITS2 type profile
						for this taxon due to its inability to identify a common set of sequences that reoccur in
						multiple samples. In this situation, a conservative ITS2 type profile will be assigned to each
						of these samples. However, by running your dataset against the remotely hosted SymPortal
						database, you enable SymPortal to search all previously submitted samples for sets of sequences
						that are also found in your samples. By this means, you may find sets of sequences that match
						your two samples, in samples already found in the database. As such, the more samples are
						contained in a database, the greater our power to resolve putative Symbiodiniaceae taxa is.
					</p>

					<div class="image round right">
						<img src="images/text_one.jpg" alt="Pic 02" />
					</div>
					<p><font color="#565656">How is my data used?</font><br/>Unpublished (and not publically available)
						data submitted to SymPortal will not be directly available to other users. However,
						immediately after submission, other users of SymPortal will be able to benefit from your data,
						just as you will be able to benefit from theirs. This is because during the SymPortal analysis
						the sets of sequences found in your samples are also searched for in all other previously
						submitted samples (as described in the Why submit my data section above). If your samples share
						ITS2 type profiles in common with other samples in the database you will be informed of this in
						your ITS2 type profile output count table. For each ITS2 type profile an ‘ITS2 type abundance
						local’ and an ‘ITS2 type abundance DB' are reported. The former, subtracted from the latter,
						will give you an indication of how many samples in the database, that were not part of your
						data set, contained the ITS2 type profile in question. No other information from non-user
						samples are returned.</p>

				</div>
			</section>

		<!-- Footer -->
			<footer id="footer">
				<div class="inner">

					<!--<h3>Get in touch for a submission link</h3>-->

					<!--<form action="#" method="post">-->

						<!--<div class="field half first">-->
							<!--<label for="name">Name</label>-->
							<!--<input name="name" id="name" type="text" placeholder="Name">-->
						<!--</div>-->
						<!--<div class="field half">-->
							<!--<label for="email">Email</label>-->
							<!--<input name="email" id="email" type="email" placeholder="Email">-->
						<!--</div>-->
						<!--<div class="field">-->
							<!--<label for="message">Message</label>-->
							<!--<textarea name="message" id="message" rows="6" placeholder="Message"></textarea>-->
						<!--</div>-->
						<!--<ul class="actions">-->
							<!--<li><input value="Send Message" class="button alt" type="submit"></li>-->
						<!--</ul>-->
					<!--</form>-->

					<div class="copyright">
						SymPortal: a novel analytical framework and platform for coral algal symbiont next-generation sequencing ITS2 profiling<br/>
						Collaborators: 	<a href="https://reefgenomics.kaust.edu.sa/Pages/Home.aspx">Reef Genomics, KAUST</a>;
										<a href="https://www.southampton.ac.uk/oes/research/facilities/corals.page">The Coral Reef Laboratory, Univerisity of Southampton</a>;
										<a href="https://nyuad.nyu.edu/en/research/centers-labs-and-projects/marine-biology-lab.html">Marine Biology Lab, NYU Abu Dhabi</a>;
										<a href="http://www.personal.psu.edu/tcl3/">The Symbiosis Ecology and Evolution Laboratory, Penn State</a>;<br/>
						Images: <a href="https://rsrc.kaust.edu.sa/Pages/Radecker.aspx">N. Raedecker</a>, <a href="https://rsrc.kaust.edu.sa/Pages/Pogoreutz.aspx">C. Pogoreutz</a>, <a href="https://rsrc.kaust.edu.sa/Pages/Gegner.aspx">H. Gegner</a>.<br/><br/>
						Contact: <a href="mailto:benjamin.hume@kaust.edu.sa?Subject=new submission enquiry">&#098;&#101;&#110;&#106;&#097;&#109;&#105;&#110;&#046;&#104;&#117;&#109;&#101;&#064;&#107;&#097;&#117;&#115;&#116;&#046;&#101;&#100;&#117;&#046;&#115;&#097;</a>
					</div>

				</div>
			</footer>

		<!-- Scripts -->
			<script src="assets/js/jquery.min.js"></script>
			<script src="assets/js/skel.min.js"></script>
			<script src="assets/js/util.js"></script>
			<script src="assets/js/main.js"></script>

	</body>
</html>